---
title: "Visualization"
author: "Tianyi"
date: "4/27/2021"
output: 
  html_document:
    toc: true
    toc_float: true
    number_sections: true
params:
    counts_matrix: "/data/subreadCounts.txt"
    sample_info: "/data/info.csv"
    cell_line: "COV434 KGN Primary"
    chemical: "DES KTZ" 
---

```{r setup, include=FALSE}
.libPaths("/Users/tili/miniconda3/envs/Cell_bulk_202011/lib/R/library")
knitr::opts_chunk$set(echo = TRUE)
```

# Cell line bulk RNAseq analysis - DES and KTZ samples

Visualization of DE genes.  

## Load required library

```{r Load dependencies}
library(DESeq2)
library(dplyr)
library(ggplot2)
library(biomaRt)
library(pheatmap)
library(VennDiagram)
library(RColorBrewer)
library(tidyverse)
```

## Data input

```{r data input}
final <- read.csv(file = "/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/tables/KTZ/sig_0_1_DE_genes_KTZ5_KGN.csv", header = T, sep = ",")
all <- read.csv(file = "/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/tables/KTZ/background_genes_shrunk_KTZ5_KGN.csv", header = T, sep = ",")

cells <- read.delim(file = "/Users/tili/Desktop/Cell_line_bulk_seq_202011/data/subreadCounts.txt")
info <- read.csv(file = "/Users/tili/Desktop/Cell_line_bulk_seq_202011/data/info.csv", header = T, sep = ";")
cells[is.na(cells)] <- 0 
```

## Volcano plot using all genes in the dataset

```{r plot volcano, fig.height = 10, fig.width = 10}
theme <- theme(text=element_text(size=14),
               axis.text.x=element_text(color="black", angle = 90),
               axis.text.y=element_text(color="black"),
               axis.ticks=element_blank(),
               strip.background = element_rect(colour=NA, fill=NA),
               plot.background = element_blank(),
               panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(),
               panel.border = element_blank(),
               axis.line.x = element_line(size=0.4),
               axis.line.y = element_line(size=0.4),
               panel.background = element_blank(),
               legend.position = "bottom")

all <- all %>% na.omit(all$pvalue)
ggplot(all, aes(x=log2FoldChange, y=-log10(pvalue))) +
    geom_point(alpha = 0.5) + xlab("Log2 fold change") + ylab("-log10 (p value)") +
    theme + ggtitle("Volcano plot") + xlim(c(-0.000001,0.000001)) +
    ggrepel::geom_text_repel(data = all %>% filter(padj < 0.1), aes(label = external_gene_name))

```

## Rename rows for count matrix

```{r rename row names for count matrix}
gene <- cells[,1] 
rownames(cells) <- gene 
cells <- cells[,-1] 

outlier <- c("COV_P26_KTZ_9", "patient3_KTZ_5")
remove_id <- match(outlier, info$sample)
info_outlier_removed <- info[-remove_id,]

genomic_idx <- match(info_outlier_removed$sample, colnames(cells))
genomic_idx
cells_ordered <- cells[ ,genomic_idx]
all(info_outlier_removed$sample == colnames(cells_ordered))
```

## Construct dds object and run DESeq analysis to obtain normalized factor for plotting

```{r construct dds object and run DESeq analysis}
dds <- DESeqDataSetFromMatrix(countData=cells_ordered, colData = info_outlier_removed, design = ~ cell_line + group) 

dds$group <- relevel(dds$group, ref = "DMSO")
dds <-DESeq(dds)
```

## Plot individual genes

```{r plot individual genes across cell lines, fig.height = 10, fig.width = 12}
plot_list <- list()
d1 <- final
for (i in 1: nrow(d1)) {
  geneID <- d1[i,2]
  gene_name <- d1[i,9]
  d2 <- plotCounts(dds, gene=geneID, intgroup=c("group", "cell_line"), normalized = T,
                  returnData=TRUE) 
  gg <- ggplot(d2, aes(x=group, y=count)) + geom_boxplot() + facet_grid(~ cell_line) +
    geom_point(size=2, alpha = 0.5) + xlab("") + ylab("normalized count") +
    theme + ggtitle(gene_name, subtitle = geneID)
  plot_list[[length(plot_list) + 1]] <- gg
}
cowplot::plot_grid(ncol = 2, nrow = 2, plotlist = plot_list) 
```

## Plot heatmap using significant DE genes

```{r heatmap with high DE genes, fig.height = 10, fig.width = 14}
d1 <- counts(dds, normalized = T) %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "geneID") %>% 
  right_join(final, by = c("geneID" = "geneID")) 
d1 <- d1[order(d1$padj, decreasing = T),]
rownames(d1) <- d1$external_gene_name
d2 <- d1[,c(2:48)]

df <- as.data.frame(colData(dds)[,c("cell_line","group")])
pheatmap(d2, cluster_rows=T, show_rownames=T,
         cluster_cols=T, annotation_col=df, fontsize_row = 10, 
         height = 20)
```

## Session infomation

```{r Session info}
sessionInfo()
```
