---
title: "Role of SCD in oocyte and GC"
author: "TL"
date: "4/26/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load dependency

```{r dependencies setup}
suppressMessages(library(Seurat))
suppressMessages(library(ggplot2))
suppressMessages(library(Matrix))
suppressMessages(library(dplyr))
suppressMessages(library(umap))
suppressMessages(library(clustree))
suppressMessages(library(ggplot2))
suppressMessages(library(cowplot))
suppressMessages(library(RColorBrewer))
suppressMessages(library(SingleCellExperiment))
suppressMessages(library(slingshot))
suppressMessages(library(stringr))
suppressMessages(library(SingleR))
suppressMessages(library(patchwork))
suppressMessages(library(biomaRt))
```



```{r data input, fig.width=15, fig.height=15}
C <- read.delim("/Users/tili/Desktop/Cell_line_bulk_seq_202011/data/oocyte_GC/count_2021.txt")
Fo <- read.delim("/Users/tili/Desktop/Cell_line_bulk_seq_202011/data/oocyte_GC/GSE107746_Folliculogenesis_FPKM.log2.txt")

F_info <- read.delim("/Users/tili/Desktop/Cell_line_bulk_seq_202011/data/oocyte_GC/SraRunTable_2021.txt", sep = ",") %>% 
  dplyr::select(Run, Developmental_stage)

C_info <- read.delim("/Users/tili/Desktop/Cell_line_bulk_seq_202011/data/oocyte_GC/SraRunTable_2018.txt", sep = ",") %>% 
  dplyr::select(Run, Cell_type)

output <- read.csv2(file = "/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/mart.csv", sep = ",") %>% 
  dplyr::select(external_gene_name, ensembl_gene_id)

get_id <- function(res.sig) {
  res.sig$geneid <- rownames(res.sig)
  final <- merge(res.sig, output, by.x = "Geneid", by.y = "ensembl_gene_id", all.x = T, all.y = FALSE)
  return(final)
}
C <- get_id(C)
C <- C[-c(1:2),]

re_name <- function(cells) {
  gene <- cells[,1]
  rownames(cells) <- gene
  cells <- cells[,-1]
  return(cells)
}

for (i in 1:nrow(C)) {
  if(i %in% which(duplicated(C$external_gene_name))) {
    if(Fo[i,ncol(C)] == "") {
      Fo[i,ncol(C)] <- paste0(C[i, 1], "1")
    } else {
          Fo[i,ncol(C)] <- paste0(C[i,ncol(C)], "1")
    }
  }
}

df <- C[!duplicated(C$external_gene_name),] %>% na.omit()
df <- df[,-1]
rownames(df) <- df$external_gene_name
df <- df[,-c(142:143)]
which(rownames(df) == "")
df <- df[-1603,]

df <- as.matrix(df)

COC <- CreateSeuratObject(counts = df, project = "COCs_2021")
C_info$Cell_type <- as.factor(C_info$Cell_type)
COC <- AddMetaData(COC, metadata = C_info$Cell_type, col.name = "cell_type")

Follicle <- CreateSeuratObject(counts = Fo, project = "COC_follicle_2018")

# Subset granulosa and oocyte clusters

alldata.int <- readRDS("/Users/tili/Desktop/scRNAseq_adult_ovary/data/Unsorted_Manuscript figures.rds")
alldata.int <- RenameIdents(alldata.int, "1_oocytes" = "Oocytes", "2_immune" = "Immune",
                            "3_gran" = "Granulosa", "4_endo" = "Endothelial", "5_pv" = "Perivascular",
                            "6_stroma" = "Stroma")

oo_GC <- subset(alldata.int, idents = c("Granulosa", "Oocytes"))


# get mitocondria gene percentage

Follicle[["percent.mt"]] <-PercentageFeatureSet(object = Follicle, pattern = "^MT-")

COC[["percent.mt"]] <-PercentageFeatureSet(object = COC, pattern = "^MT-")


# plot violin plot to decide filtering factors

VlnPlot(Follicle, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

wrap_plots(plotlist = qc_list, ncol = 2, nrow = 3)

Follicle <- subset(Follicle, subset = nFeature_RNA > 200 & nFeature_RNA < 20000 & percent.mt < 20)


```


## Normalization 

```{r normalize}

Follicle <- NormalizeData(Follicle, normalization.method = "LogNormalize", scale.factor = 10000)
Follicle <- FindVariableFeatures(Follicle, selection.method = "vst", nfeatures = 2000)

```

## Cell scoring for all patinets

```{r merge and check if C759 can be combined}

s_genes <- cc.genes$s.genes
g2m_genes <- cc.genes$g2m.genes

Follicle <- CellCycleScoring(Follicle, s.features = s_genes, g2m.features = g2m_genes, set.ident = T)

```

## Scaling

```{r Scaling}

# features <- rownames(Follicle)
Follicle <- ScaleData(Follicle)
Follicle <- RunPCA(Follicle, features = VariableFeatures(object = Follicle))

```


## Integration

```{r integration}

cell_type <- oo_GC@active.ident %>% as.data.frame() %>% 
  'colnames<-' ("cell_type")
oo_GC <- AddMetaData(oo_GC, metadata = cell_type$cell_type, col.name = "cell_type")
oo_GC[["seurat_clusters"]] <- NULL
control_int <- list(Follicle, oo_GC)

# Find anchor and integrate dataset
features <- SelectIntegrationFeatures(control_int)
control_int <- lapply(X = control_int, FUN = function(x) {
    x <- ScaleData(x, features = features)
    x <- RunPCA(x, features = features)
})

anchors <- FindIntegrationAnchors(control_int, reduction = "cca", dims = 1:50)
control_int <- IntegrateData(anchorset = anchors, dims = 1:50, k.weight = 90)

# Scale and perform dimensional reduction
control_int <- ScaleData(control_int)
control_int <- FindVariableFeatures(control_int, selection.method = "vst", nfeatures = 2000)
control_int <- RunPCA(control_int, npcs = 50, features = VariableFeatures(control_int))
DimPlot(control_int, reduction = "pca", group.by = "cell_type")

DimHeatmap(control_int, cells = 500, dims = 1:20, balanced = T)
ElbowPlot(control_int, ndims = 50)

control_int <- RunPCA(control_int, npcs = 13)
control_int <- RunUMAP(control_int, dims = 1:13, reduction = "pca", seed.use = 100)
DimPlot(control_int, reduction = "umap", group.by = "cell_type")

DefaultAssay(control_int) <- "integrated"
for (res in seq(0, 1, by = 0.1)) {
  control_int <- FindClusters(object = control_int, resolution = res)
}
control_int@meta.data
clustree(control_int, prefix = "integrated_snn_res.", exprs = c("scale.data"))

control_int <- FindNeighbors(object = control_int, reduction = "pca", dims = 1:13)
control_int <- FindClusters(control_int, resolution = 0.2)
control_int <- RunUMAP(control_int, reduction = "pca", dims = 1:13, reduction.name = "umap", seed.use = 100)
DimPlot(control_int, reduction = "umap")

p1 <- plot_umap(control_int, "Phase")
p3 <- plot_umap(control_int, "old.ident")
p2 <- plot_umap(control_int, "ident")
p4 <- plot_umap(control_int, "percent.mt")
p5 <- plot_umap(control_int, "nCount_RNA")
p6 <- plot_umap(control_int, "nFeature_RNA")

p1 + p2 + p3 + p4 + p5 + p6

DotPlot(control_int, features = "SCD")
```

## SCT transformation

```{r SCT transformation - Data integration, fig.width=15, fig.height=15}

control_int <- lapply(control_int, FUN = function(x) {
  x <- NormalizeData(x, normalization.method = "LogNormalize", scale.factor = 10000)
})

s_genes <- cc.genes$s.genes
g2m_genes <- cc.genes$g2m.genes

control_int <- lapply(control_int, FUN = function(x) {
  x <- CellCycleScoring(x, s.features = s_genes, g2m.features = g2m_genes, set.ident = T)
})


control_int <- lapply(control_int, FUN = SCTransform) # vars.to.regress argument can be added in the functions later if needed
features <- SelectIntegrationFeatures(control_int, nfeatures = 2000)
control_int <- PrepSCTIntegration(control_int, anchor.features = features)

# Find anchor and perform SCT integration
sct_anchors <- FindIntegrationAnchors(control_int, normalization.method = "SCT", anchor.features = features)
control_int_sct <- IntegrateData(anchorset = sct_anchors, normalization.method = "SCT", k.weight = 90)

control_int_sct <- SCTransform(control_int_sct)

DefaultAssay(control_int_sct) <- "integrated"
# Perform dimensional reduction (ScaleData is not needed for SCT transformation)
control_int_sct <- RunPCA(control_int_sct, npcs = 13)
control_int_sct <- RunUMAP(control_int_sct, dims = 1:13, reduction = "pca", seed.use = 100)
DimPlot(control_int_sct, reduction = "umap", group.by = "cell_type")

control_int_sct <- FindNeighbors(object = control_int_sct, reduction = "pca", dims = 1:13)

for (res in seq(0, 1, by = 0.1)) {
  control_int_sct <- FindClusters(object = control_int_sct, resolution = res)
}
control_int_sct@meta.data
clustree(control_int_sct, prefix = "integrated_snn_res.", exprs = c("scale.data"))

control_int_sct <- FindClusters(control_int_sct, resolution = 0.1)
control_int_sct <- RunUMAP(control_int_sct, reduction = "pca", dims = 1:13, reduction.name = "umap", seed.use = 100)
DimPlot(control_int_sct, reduction = "umap")

control_int_sct <- RenameIdents(control_int_sct, "0" = "Granulosa", "1" = "Oocytes")
DimPlot(control_int_sct, reduction = "umap")
DimPlot(control_int_sct, reduction = "umap", group.by = "orig.ident")
DotPlot(control_int_sct, features = "SCD")

plot_umap <- function(object, x) {
  DimPlot(object = object, reduction = "umap", group.by = x, pt.size=1, 
    repel = F
    ) + NoLegend()
}

p1 <- plot_umap(control_int_sct, "Phase")
p3 <- plot_umap(control_int_sct, "old.ident")
p2 <- plot_umap(control_int_sct, "ident")
p4 <- plot_umap(control_int_sct, "percent.mt")

p1 + p2 + p3 + p4

# Generate count and metadata for cellphonedb analysis
embed <- control_int_sct@assays$SCT@data
write.table(embed, file = "/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/cpdb_count.txt")

meta <- control_int_sct@meta.data
write.table(meta, file = "/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/cpdb_meta.txt")

save(oo_GC, Follicle, file = "/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/oocyte_GC/seurat_object_oo_GC_COC.Rdata")

```

# Visualization of cellphonedb results

```{r cellphonedb results}
sig_mean <- read.delim("/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/cellphonedb/out/significant_means.txt")
p <- read.delim("/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/cellphonedb/out/pvalues.txt")

enrich <- p[(p$Granulosa.Oocytes < 0.05 | p$Oocytes.Granulosa < 0.05),]
enrich <- enrich[(enrich$Granulosa.Oocytes != 0 & enrich$Oocytes.Granulosa != 0),]

theme <- theme(text=element_text(size=15),
               axis.text.x=element_text(color="black", angle = 90),
               axis.text.y=element_text(color="black"),
               axis.ticks=element_blank(),
               strip.background = element_rect(colour=NA, fill=NA),
               plot.background = element_blank(),
               panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(),
               panel.border = element_blank(),
               axis.line.x = element_line(size=0.4),
               axis.line.y = element_line(size=0.4),
               panel.background = element_blank(),
               legend.position = "bottom")

enrich1 <- enrich %>% dplyr::select(interacting_pair, Granulosa.Oocytes, Oocytes.Granulosa) %>% 
  tidyr::pivot_longer(c(Granulosa.Oocytes, Oocytes.Granulosa), names_to = "variable", values_to = "value")

ggplot(enrich1, aes(y=interacting_pair, x=variable, col = variable)) + geom_point(aes(size = value)) +
  theme + coord_flip()

gene <- c(enrich$gene_a, enrich$gene_b) %>% as.data.frame() %>% 'colnames<-' ("genes")
gene <- gene[!(gene$genes == ""),]
```

# Overlapped ligands/receptors with rat data DEGs

```{r overlap with rats DEGs}

overlap <- intersect(gene, toupper(set$external_gene_name))
d1 <- enrich[match(overlap, enrich$gene_a),]
d2 <- enrich[match(overlap, enrich$gene_b),]

d3 <- rbind(d1, d2) %>% na.omit()

# expression in rat DEGs list
for (i in 1:18) {
  print(names(final[i]))
  print(intersect(overlap, toupper(final[[i]]$external_gene_name)))
}

```



## Find cluster markers

```{r find cluster markers}
cluster_markers <- FindAllMarkers(child$C973, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25) %>% 
  group_by(cluster) %>% top_n(n=10, wt=avg_log2FC)

cluster_markers %>%
    group_by(old.ident) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10
DoHeatmap(child$C973, features = top10$gene) + NoLegend()
```

# Batch of follicles in human and rat (adult)

```{r similarities between human and rat follicle batch}
hg <- read.csv2("/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/oocyte_GC/human_adult_follicle_sequencing/adult_counts2.csv")
hg_info <- read.csv2("/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/oocyte_GC/human_adult_follicle_sequencing/adult_samples2.csv")

rownames(hg) <- hg$genes
hg <- hg[,-1]

dds_hg <- DESeqDataSetFromMatrix(countData=hg, colData = hg_info, design = ~ stage)
r_hg <- edgeR::cpm(counts(dds_hg), normalized.lib.sizes = TRUE, log = T)
r_hg <- as.data.frame(r_hg)
r_hg$id <- rownames(r_hg)

r <- edgeR::cpm(counts(dataset[["dds_adult_DES"]]), normalized.lib.sizes = T, log = T)
r <- as.data.frame(r) %>% get_id()
r$external_gene_name <- toupper(r$external_gene_name)
r_rat <- r[,-c(1,20, 22:23)]

all <- merge(r_hg, r_rat, by.x = "id", by.y = "external_gene_name", all.x=FALSE, all.y =T)

ful <- all 
ful[is.na(ful)] <- 0
ful[ful$id == "",]$id <- rownames(ful[ful$id == "",])
ful <- ful[!duplicated(ful$id),]
rownames(ful) <- ful$id
ful <- ful[,-1]
ful <- ful[rowSums(ful) >1,]

pca <- prcomp(t(ful))
info_rat <- read.csv2("/Users/tili/Desktop/FREIA_rat_202007/data/processed/info_adult_DES.csv", sep = ",") %>% 
  dplyr::select(name, treatment)
info_hu <- hg_info %>% dplyr::select(label, stage) %>% 'colnames<-' (c("name", "treatment"))
info <- rbind(info_hu, info_rat)
info$species <- factor(c(rep("human", 60), rep("rat", 18)))

df <- cbind(info, pca$x)
contri <- round(summary(pca)$importance[2,] * 100, 2)
ggplot(df) + geom_point(aes(x=PC1, y = PC2, col = species), size = 3) + theme + 
      ggtitle("PCA (cpm)") + scale_color_manual(values = c("#8dd3c7","#fb8072", "#bebada", "blue")) +
      xlab(sprintf("PC1 (%s%%)", contri[1])) + ylab(sprintf("PC2 (%s%%)", contri[2]))
```



################################################# Trajectory analysis ############################################

```{r slingshot trajectory analysis}
# get normalized count matrix
counts <- ovar_child1@assays$RNA@data
sim <- SingleCellExperiment(assays = List(counts = counts))

# Reduced dimension (i.e. obtained by PCA, ICA, diffusion maps etcs)
rd <- as.matrix(ovar_child1[['umap']]@cell.embeddings)
dim(rd)

# cluster labels generated by k means clustering
cl <- as.data.frame(ovar_child1@active.ident)
length(cl)

cl$clu <- cl$`ovar_child1@active.ident` %>%  str_replace_all("0","Oocytes") %>% 
  str_replace_all("5","Immune") %>% 
  str_replace_all("2","Granulosa") %>% 
  str_replace_all("3","Unknown") %>% 
  str_replace_all("4","Perivascular") %>% 
  str_replace_all("1","unknown")

cluster <- as.vector(cl$clu)
length(cluster)

```

```{r dimensional reduction}
reducedDims(sim) <- SimpleList(PCA = ovar_child1@reductions$pca@cell.embeddings[,1:2],
                               UMAP = ovar_child1@reductions$umap@cell.embeddings[,1:2])
colData(sim)$kmeans <- cluster

```

```{r slingshot analysis}
sim <- slingshot(sim, clusterLabels = "kmeans", reducedDim = "UMAP")

colors <- colorRampPalette(brewer.pal(11,'Spectral')[-6])(100)
plotcol <- colors[cut(sim$slingPseudotime_1, breaks=100)]

plot(reducedDims(sim)$UMAP, col = plotcol, pch=16, asp = 1, xlim = c(-13,13))
lines(SlingshotDataSet(sim), lwd=2, col='black')

plot(reducedDims(sim)$UMAP, pch=16, asp = 1)
lines(SlingshotDataSet(sim), lwd=2, type = 'lineages', col = 'blue')

```

```{r identifyting global lineage structure}
lin1 <- getLineages(rd, cluster, start.clus = "0")

plot(rd, asp = 0.8, pch = 16, col = cl$`ovar_child1@active.ident`)
lines(lin1, lwd = 3, col = 'black')

lin2 <- getLineages(rd, cl, start.clus= '1', end.clus = '2')

plot(rd, col = cl, asp = 1, pch = 0.5)
lines(lin2, lwd = 3, col = 'blue', show.constraints = TRUE)

map <- cbind(data.frame(rd), data.frame(cl))
ggplot(map, aes(UMAP_1, UMAP_2, col = cl)) + geom_point(size=0.5) + theme(text=element_text(size=14),
               axis.text.x=element_text(color="black", angle = 90),
               axis.text.y=element_text(color="black"),
               axis.ticks=element_blank(),
               strip.background = element_rect(colour=NA, fill=NA),
               plot.background = element_blank(),
               panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(),
               panel.border = element_blank(),
               axis.line.x = element_line(size=0.4),
               axis.line.y = element_line(size=0.4),
               panel.background = element_blank())  
lin2@lineages
```

```{r monocle trajectory analysis}
saveRDS(ovar_child1, file = "/Users/tili/Desktop/scRNAseq_child_ovary/data/seurat_v4_res_0_1_PC12.rds")
ovar_child1$assigned_celltype <- Idents(ovar_child1)
gr <- ovar_child1[, ovar_child1$assigned_celltype %in% c("2")]
gr.cds <- SeuratWrappers::as.cell_data_set(gr)
gr.cds <- cluster_cells(cds = gr.cds, reduction_method = "UMAP")
gr.cds <- learn_graph(gr.cds, use_partition = T)
```


```{r session info}
session_info()
```

